sql: support spilling to disk for bufferNode

yuzefovich opened the pull request:
This commit refactors several `planNode`s that need to buffer rows to
use a disk-backed row container instead of pure in-memory one. In order
to achieve that a couple of light wrappers are introduced on top of the
corresponding container and an iterator over it.

Still, one - probably important - case is not fixed properly: namely, if
a subquery is executed in AllRows or AllRowsNormalized mode, then we
first buffer the rows into the disk-backed container only to materialize
it later into a single tuple. Addressing this is left as a TODO.

Fixes: #62301.
Fixes: #62674.

Release note (sql change): CockroachDB now should be more stable when
executing queries with subqueries producing many rows (previously we
could OOM crash and now we will use the temporary disk storage).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63900)
<!-- Reviewable:end -->

yuzefovich committed:
sql: use disk-backed container in bufferNode

This commit refactors several `planNode`s that need to buffer rows to
use a disk-backed row container instead of pure in-memory one. In order
to achieve that a couple of light wrappers are introduced on top of the
corresponding container and an iterator over it.

Still, one - probably important - case is not fixed properly: namely, if
a subquery is executed in AllRows or AllRowsNormalized mode, then we
first buffer the rows into the disk-backed container only to materialize
it later into a single tuple. Addressing this is left as a TODO.

Release note (sql change): CockroachDB now should be more stable when
executing queries with subqueries producing many rows (previously we
could OOM crash and now we will use the temporary disk storage).

yuzefovich commented:
Kicked off 20 runs of `tpcdsvec` (re #62301), and they finished successfully with this commit.

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 11 of 11 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63900#-:-MYlqL714y9q0JxABIoG:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/apply_join.go, line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63900#-MYjzpV53KnqlhxSxZbE:-MYjzpV53KnqlhxSxZbF:b624yrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae7992fd054bcd798168a501a9c3eaa478482069/pkg/sql/apply_join.go#L124)):*
> ```Go
> 			for {
> 				var rrow tree.Datums
> 				if len(a.rightTypes) != 0 {
> ```

how will we know when the rows are exhausted if there are no columns?
___
*[pkg/sql/create_stats.go, line 522 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63900#-MYloESqD5Nga2iLZIkm:-MYloESqD5Nga2iLZIkn:b8774wh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae7992fd054bcd798168a501a9c3eaa478482069/pkg/sql/create_stats.go#L522)):*
> ```Go
> 		planCtx := dsp.NewPlanningCtx(ctx, evalCtx, nil /* planner */, txn, true /* distribute */)
> 		if err := dsp.planAndRunCreateStats(
> 			ctx, evalCtx, planCtx, txn, r.job, NewRowResultWriter(nil /* rowContainer */),
> ```

how does this change without a row container? Shouldn't we pass a rowContainerHelper?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63900#-:-MYqeYOXAfnHwjxg5Dh1:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/apply_join.go, line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63900#-MYjzpV53KnqlhxSxZbE:-MYpRtZaAVNQ1NdjlGHg:b-nwtak4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae7992fd054bcd798168a501a9c3eaa478482069/pkg/sql/apply_join.go#L124)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

how will we know when the rows are exhausted if there are no columns?
</blockquote></details>

Nice catch! In this scenario previously we would have entered an infinite loop. I added a test and fixed the problem (this also required updating the mem row container code slightly).
___
*[pkg/sql/create_stats.go, line 522 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63900#-MYloESqD5Nga2iLZIkm:-MYpQJwrBgEoQ8A3he-o:b-ck17al) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae7992fd054bcd798168a501a9c3eaa478482069/pkg/sql/create_stats.go#L522)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

how does this change without a row container? Shouldn't we pass a rowContainerHelper?
</blockquote></details>

`sampleAggregator` can only produce metadata objects, so we don't need an actual row container here. Added a comment.


<!-- Sent from Reviewable.io -->
